addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/AbstractJeeLinkConnection.java

```diff
@@ -0,0 +1,131 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.jeelink.handler;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.ConnectException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.openhab.binding.jeelink.config.JeeLinkConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract base class for a connection to a JeeLink.
+ * Manages ReadingListeners, finds out the sketch name and allows to propagate read lines.
+ *
+ * @author Volker Bier - Initial contribution
+ */
+public abstract class AbstractJeeLinkConnection implements JeeLinkConnection {
+    private final Logger logger = LoggerFactory.getLogger(JeeLinkSerialConnection.class);
+
+    protected final ArrayList<JeeLinkReadingConverter<?>> inputHandlers = new ArrayList<>();
+    protected String sketchName = null;
+    protected String port;
+
+    private String[] initCommands;
+
+    public AbstractJeeLinkConnection(String port) {
+        this.port = port;
+    }
+
+    @Override
+    public String getPort() {
+        return port;
+    }
+
+    @Override
+    public void addReadingConverter(JeeLinkReadingConverter<?> listener) {
+        synchronized (inputHandlers) {
+            if (!inputHandlers.contains(listener)) {
+                logger.debug("Added listener {}.", listener);
+                inputHandlers.add(listener);
+            }
+        }
+    }
+
+    @Override
+    public void removeReadingConverters() {
+        synchronized (inputHandlers) {
+            inputHandlers.clear();
+        }
+    }
+
+    public void propagateLine(String line) {
+        logger.trace("Read line from port {}: {}", port, line);
+
+        if (sketchName == null) {
+            Matcher matcher = Pattern.compile("\\[(\\w+).*\\]").matcher(line);
+            boolean matches = matcher.matches();
+            if (matches) {
+                sketchName = matcher.group(1);
+                logger.debug("Sketch name found for port {}: {}", port, sketchName);
+
+                initializeDevice();
+            }
+        }
+
+        synchronized (inputHandlers) {
+            for (JeeLinkReadingConverter<?> l : inputHandlers) {
+                l.handleInput(line);
+            }
+        }
+    }
+
+    @Override
+    public void setInitCommands(String initCommands) {
+        if (initCommands != null && !initCommands.trim().isEmpty()) {
+            this.initCommands = initCommands.split(";");
+        }
+    }
+
+    private void initializeDevice() {
+        try {
+            if (initCommands != null) {
+                logger.debug("Initializing device on port {} with commands {} ", port, Arrays.toString(initCommands));
```

Comment:
The array to String conversion will always be performed (also if you are not running in debug), you could put it in a check for logger.isDebugEnabled. I don't know how often we execute this line, most likely the impact is really minor.